include/boost/math/differentiation/autodiff.hpp

```diff
@@ -0,0 +1,1882 @@
+    // ca < cr | bool | Less than comparison.
+    template<typename RealType2,size_t Order2>
+    friend bool operator<(const typename dimension<RealType2,Order2>::root_type&, const dimension<RealType2,Order2>&);
+    // cr > cr2 | bool | Greater than comparison.
+    template<typename RealType2,size_t Order2>
+    bool operator>(const dimension<RealType2,Order2>&) const;
+    // cr > ca | bool | Greater than comparison.
+    bool operator>(const root_type&) const;
+    // ca > cr | bool | Greater than comparison.
+    template<typename RealType2,size_t Order2>
+    friend bool operator>(const typename dimension<RealType2,Order2>::root_type&, const dimension<RealType2,Order2>&);
+
+    // Will throw std::out_of_range if Order < order.
+    template<typename... Orders>
+    typename type_at<RealType,sizeof...(Orders)>::type at(size_t order, Orders... orders) const;
+    template<typename... Orders>
+    typename type_at<RealType,sizeof...(Orders)-1>::type derivative(Orders... orders) const;
+
+    dimension<RealType,Order> inverse() const; // Multiplicative inverse.
+
+    static constexpr size_t depth(); // Number of nested std::array<RealType,Order>.
+    static constexpr size_t order_sum();
+
+    explicit operator root_type() const; // Must be explicit, otherwise overloaded operators are ambiguous.
+    dimension<RealType,Order>& set_root(const root_type&);
+
+    // Use when function returns derivatives.
+    dimension<RealType,Order> apply(const std::function<root_type(size_t)>&) const;
+    // Use when function returns derivative(i)/factorial(i) (slightly more efficient than apply().)
+    dimension<RealType,Order> apply_with_factorials(const std::function<root_type(size_t)>&) const;
+    // Same as apply() but uses horner method. May be more accurate in some cases but not as good with inf derivatives.
+    dimension<RealType,Order> apply_with_horner(const std::function<root_type(size_t)>&) const;
+    // Same as apply_with_factorials() but uses horner method.
+    dimension<RealType,Order> apply_with_horner_factorials(const std::function<root_type(size_t)>&) const;
+
+private:
+    RealType epsilon_inner_product(size_t z0, size_t isum0, size_t m0,
+        const dimension<RealType,Order>& cr, size_t z1, size_t isum1, size_t m1, size_t j) const;
+    dimension<RealType,Order> epsilon_multiply(size_t z0, size_t isum0,
+        const dimension<RealType,Order>& cr, size_t z1, size_t isum1) const;
+    dimension<RealType,Order> epsilon_multiply(size_t z0, size_t isum0, const root_type& ca) const;
+    dimension<RealType,Order> inverse_apply() const;
+    dimension<RealType,Order>& multiply_assign_by_root_type(bool is_root, const root_type&);
+
+    template<typename RealType2,size_t Orders2>
+    friend class dimension;
+    template<typename RealType2,size_t Order2>
+    friend std::ostream& operator<<(std::ostream&, const dimension<RealType2,Order2>&);
+
+// C++11 Compatibility
+#ifdef BOOST_NO_CXX17_IF_CONSTEXPR
```

Comment:
I was originally afraid of that as well but a volunteer from the mailing list did some pretty clean C++11 compatibility integration, without having to use SFINAE. Also since I'll be moving things into `boost::math::differentiation::detail` I can replace the static constexpr class methods which were giving MSVC2015 some problems, which should make it "more" C++11 compatible. I would also hate to have to undo his contribution. I think the C++11 additions/alternative are pretty straight-forward.

Reply:
Done. Compiles now with MSVC 2015. https://ci.appveyor.com/project/pulver/autodiff/builds/22098322